### PR TITLE
feat: add `env package` command

### PR DESCRIPTION
**Pull request by efekarakus**
Related #3522

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/cli/env_package.go [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+
+	name, err := o.sel.LocalEnvironment("Select an environment manifest from your workspace", "")
+	if err != nil {
+		return fmt.Errorf("select environment: %w", err)
+	}
+	o.envName = name
+	return nil
+}
+
+func (o *packageEnvOpts) environmentManifest() (*manifest.Environment, error) {
```
Comment (Lou1415926) [inline internal/pkg/cli/env_package.go]: Do you think it's worthy uniting this method with the following? https://github.com/aws/copilot-cli/blob/74d034755279b4d28b3c6f4b61c9a34aa2e6513b/internal/pkg/cli/env_deploy.go#L138

File: internal/pkg/cli/env_package.go [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+func (o *packageEnvOpts) setWriters() error {
+	if o.outputDir == "" {
+		return nil
+	}
+	if err := o.fs.MkdirAll(o.outputDir, 0755); err != nil {
+		return fmt.Errorf("create directory %q: %w", o.outputDir, err)
+	}
+
+	path := filepath.Join(o.outputDir, fmt.Sprintf("env-%s.yml", o.envName))
```
Comment (Lou1415926) [inline internal/pkg/cli/env_package.go]: An extreme edge case would be having a workload called "env". Running `svc package --name env --env test` would write the output to `env-test.params.json`  - because it's formatted as `<svc>-<env>.params.json` - which collides with this naming scheme. 

I don't think we need to worry about this too much though but just to call it out!

File: internal/pkg/cli/env_package.go [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+	packageEnvVars
+
+	// Dependencies.
+	cfgStore     store
+	ws           wsEnvironmentReader
+	sel          wsEnvironmentSelector
+	caller       identityService
+	fs           afero.Fs
+	tplWriter    io.WriteCloser
+	paramsWriter io.WriteCloser
```
Comment (Lou1415926) [inline internal/pkg/cli/env_package.go:59]: We can use `WriteCloser` for `svc package`'s writers as well! In a later refactor

**Review by Lou1415926 — APPROVED** [review]

File: internal/pkg/cli/env_package.go (hunk 2 above)
Comment (iamhopaul123) [reply internal/pkg/cli/env_package.go]: To avoid this edge case can we simply change the env artifacts naming pattern to `env_%s.yaml`?

File: internal/pkg/cli/env_package.go [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+
+// buildEnvPkgCmd builds the command for printing an environment CloudFormation stack configuration.
+func buildEnvPkgCmd() *cobra.Command {
+	vars := packageEnvVars{}
+	cmd := &cobra.Command{
+		Use:   "package",
+		Short: "Prints the AWS CloudFormation template of an environment.",
+		Long:  `Prints the CloudFormation stack template and configuration used to deploy an environment.`,
+		Example: `
+  Print the CloudFormation template for the "prod" environment.
```
Comment (iamhopaul123) [inline internal/pkg/cli/env_package.go]: nit: should we be consistent about "print" or "prints"? Maybe
```suggestion
		Short: "Print the AWS CloudFormation template of an environment.",
		Long:  `Print the CloudFormation stack template and configuration used to deploy an environment.`,
		Example: `
  Print the CloudFormation template for the "prod" environment.
```

File: internal/pkg/cli/interfaces.go [hunk 5]
```diff
@@ -658,3 +658,8 @@ type envDeployer interface {
 	DeployEnvironment(in *clideploy.DeployEnvironmentInput) error
 	UploadArtifacts() (map[string]string, error)
 }
+
+type envPackager interface {
+	GenerateCloudFormationTemplate(in *clideploy.DeployEnvironmentInput) (*clideploy.GenerateCloudFormationTemplateOutput, error)
```
Comment (iamhopaul123) [inline internal/pkg/cli/interfaces.go:667]: How about
```suggestion
type packager interface {
	GenerateCloudFormationTemplate(in *clideploy.GenerateCloudFormationTemplateInput) (*clideploy.GenerateCloudFormationTemplateOutput, error)
```
💭

File: cmd/copilot/template/macro.go [hunk 6]
```diff
@@ -40,12 +40,29 @@ func h2(text string) string {
...
 	lines := strings.Split(text, "\n")
+
+	var startCodeBlockIdx, codeBlockPadding int
 	for i, line := range lines {
 		if strings.HasPrefix(strings.TrimSpace(line), "/code ") {
 			codeIndex := strings.Index(line, "/code ")
 			lines[i] = line[:codeIndex] +
 				termcolor.HighlightCode(strings.ReplaceAll(line[codeIndex:], "/code ", ""))
 		}
+		if strings.HasPrefix(strings.TrimSpace(line), "/startcodeblock") {
```
Comment (iamhopaul123) [inline cmd/copilot/template/macro.go:51]: This is awesome!

File: internal/pkg/cli/env_package.go (hunk 2 above)
Comment (Lou1415926) [reply internal/pkg/cli/env_package.go]: then it may have 
```
/infrastructure
   /svc-env.yaml
   /env_test.yaml
```
The inconsistency on the usage of `-` and `_` is a bit bothering 😓

Comment (iamhopaul123) [reply internal/pkg/cli/env_package.go]: Right right but it seems like we need to choose from two poisons

Comment (efekarakus) [reply internal/pkg/cli/env_package.go]: what do you think of 
`./infrastructure/environments/env-test.yaml`

we can then generate the pipeline manifest with this path in mind

Comment (iamhopaul123) [reply internal/pkg/cli/env_package.go]: yea but then the output dir isn't really the one they specify, which is not consistent with the workload behavior. And could end up being `./infra/env/environments/env-test.yaml` if they specify the output dir to be `./infra/env/`. I would prefer to leave as is...

File: internal/pkg/cli/env_package.go (hunk 1 above)
Comment (efekarakus) [reply internal/pkg/cli/env_package.go]: Done!

File: internal/pkg/cli/env_package.go (hunk 2 above)
Comment (efekarakus) [reply internal/pkg/cli/env_package.go]: Renamed the file format to `%s.env.yml` and `%s.env.params.json`

File: internal/pkg/cli/env_package.go (hunk 3 above)
Comment (efekarakus) [reply internal/pkg/cli/env_package.go:59]: Done! updated in this pr

File: internal/pkg/cli/env_package.go (hunk 4 above)
Comment (efekarakus) [reply internal/pkg/cli/env_package.go]: Done!

File: internal/pkg/cli/interfaces.go (hunk 5 above)
Comment (efekarakus) [reply internal/pkg/cli/interfaces.go:667]: I couldn't really do that because the template inputs for environments and workloads are pretty different :rip:

**Review by Lou1415926 — APPROVED** [review]

**Review by iamhopaul123 — APPROVED** [review]
